### PR TITLE
Added vision training code

### LINKED ISSUE #13: Training and Evaluation Pipeline for RC-Net
Scripts for training and validating the region classifier network after data augmentation (#10 )

### LINKED ISSUE #14: Training and Evaluation Pipeline for LD-Net
Develop scripts for training the Landmark Detection Models (YOLOv8) on the augmented data (#10)

### LINKED ISSUE #17: Saliency maps
Need to integrate the saliency map work done by Kyle and the argus-1 people.

### LINKED ISSUE #56: Change RCNet from Single-label classifier to Multi-label classifier
Currently, both training and inference assume that the input image belongs to one of the salient regions. However, if the model receives an image outside these regions, it should output a multi-hot encoding of all zeros.

### PR DESCRIPTION
**Pull request by ArvindCar**
Fixes #13  
Fixes #14  
Fixes #17  
Fixes #56

### CODE REVIEW

File: VisionTrainingGround/run_YOLO.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
Comment (amaarquadri) [inline VisionTrainingGround/run_YOLO.sh:1]: It should be clear from the file name that this is for training the LDs. Maybe rename to `train_LD_YOLO.sh`?

**Review by amaarquadri — CHANGES_REQUESTED** [review]
I just read the main README, the bash scripts it linked to, and the module docstrings for the python scripts those linked to. I think its fine to merge this without a detailed code review as long as we understand the high-level process and can run the entire pipeline.

The main change that I would like to see is a consolidation of the 3 earth engine downloading scripts.

File: VisionTrainingGround/run_YOLO.sh (hunk 1 above)
Comment (ArvindCar) [reply VisionTrainingGround/run_YOLO.sh:1]: Changed

File: VisionTrainingGround/LD/train_yolo.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import argparse
```
[the remark this answers is not in the document]
Comment (ArvindCar) [reply VisionTrainingGround/LD/train_yolo.py:1]: Added

File: VisionTrainingGround/DataPipeline/prepare_yolo_data.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (ArvindCar) [reply VisionTrainingGround/DataPipeline/prepare_yolo_data.py:1]: Added

File: VisionTrainingGround/run_downloader.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    s) SCALE="$OPTARG" ;;
+    h) echo "Usage: $0 [options]" ;;
+    \?) echo "Invalid option: -$OPTARG" ;;
+  esac
+done
+
+# Path to python bin may be adjusted as needed
+for KEY in "${KEYS[@]}"; do
+  # Download dataset
+  python ./DataPipeline/earthenginedl.py --bounds $BOUNDS --idate "$IDATE" --fdate "$FDATE" --landsat $LANDSAT --grid_key "$KEY" --region "$KEY" --maxims $MAXIMS --scale $SCALE --outpath "$BASE_OUTPATH/$KEY"
```
[the remark this answers is not in the document]
Comment (ArvindCar) [reply VisionTrainingGround/run_downloader.sh:30]: Agreed, removed this script

File: VisionTrainingGround/run_downloader.sh [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+# Default values for configurations
+BOUNDS="-84 24 -78 32"
+IDATE="2020-05-01"
+FDATE="2023-12-31"
+LANDSAT=8
+MAXIMS=500
+SCALE=328
+BASE_OUTPATH="Landsat_Data"
+KEYS=("53L") # Add or remove keys as needed
```
[the remark this answers is not in the document]
Comment (ArvindCar) [reply VisionTrainingGround/run_downloader.sh:11]: Removed this script

File: VisionTrainingGround/DataPipeline/saliencymap2boxes.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            tl_lon, tl_lat = dataset.xy(y, x)  # Top-left
+            br_lon, br_lat = dataset.xy(y + h, x + w)  # Bottom-right
+            # Calculate centroid coordinates
+            centroid_lon = (tl_lon + br_lon) / 2
+            centroid_lat = (tl_lat + br_lat) / 2
+            outboxes.append([centroid_lon, centroid_lat, tl_lon, tl_lat, br_lon, br_lat])
+
+        # Save as .npy
+        outboxes = np.array(outboxes)
+        np.save(input_path + '/' + key + '_outboxes.npy', outboxes)
```
[the remark this answers is not in the document]
Comment (ArvindCar) [reply VisionTrainingGround/DataPipeline/saliencymap2boxes.py:85]: Changed

File: VisionTrainingGround/Salient_Region_Analysis/getMGRS.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+"""
```
Comment (amaarquadri) [inline VisionTrainingGround/Salient_Region_Analysis/getMGRS.py:1]: Is this used anywhere? I believe this is a helper module for the EEDL scripts and can now be removed.

File: VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy
Comment (amaarquadri) [inline VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy:1]: It seems to me like the contents of `/Salient_Region_Analysis/landmarks` is mostly just results from running the code? We probably don't need to have all of this on GitHub.

**Review by amaarquadri — CHANGES_REQUESTED** [review]

File: VisionTrainingGround/Salient_Region_Analysis/getMGRS.py (hunk 7 above)
Comment (ArvindCar) [reply VisionTrainingGround/Salient_Region_Analysis/getMGRS.py:1]: It's being used to make the YOLO dataset. I don't understand exactly how yet, but it's being used in saliencymap.py (to merge saliency maps of a given region I think)

File: VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy
Comment (ArvindCar) [reply VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy:1]: A lot of what is in `/Salient_Region_Analysis` are results from running different scripts in that folder. I just left them as is because that's how Kyle's repo was, should I remove them all? The only true "inputs" might be the blue marble folders, everything else is derived from them.

Comment (fredmarkus) [reply VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy:1]: imo anything that can be created from scripts/inputs should not be stored. We can described in the README how to generate stuff like this, but we should aim to ship something that is as small as possible.

File: VisionTrainingGround/Salient_Region_Analysis/getMGRS.py (hunk 7 above)
Comment (ArvindCar) [reply VisionTrainingGround/Salient_Region_Analysis/getMGRS.py:1]: It's used to consolidate the region maps, its used for salient region analysis, preparing the YOLO dataset, and in your eedl script. This should probably be imported as a package though, right now there are three identical versions of this script in 3 different folders

File: VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy
Comment (ArvindCar) [reply VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy:1]: Fair point, I've removed landmarks

Comment (ArvindCar) [reply VisionTrainingGround/Salient_Region_Analysis/landmarks/10S_landmarksf30_opts.npy:1]: I realized there were a bunch of images/csv that were also outputs, not inputs to the pipeline, they've been removed as well. Whatever is left is either an input to the saliency code or for the README

File: VisionTrainingGround/run_downloader.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArvindCar) [reply VisionTrainingGround/run_downloader.sh:30]: My bad I thought I removed this, I'll just remove it before the next PR when I push the new models. I don't think there needs to be something else in place of this though, since the image source won't be directly from EE once the sim works right? I'll just update the README to mention where the downloaded images need to go, would that be fine?

[the remark this answers is not in the document]
Comment (amaarquadri) [reply VisionTrainingGround/run_downloader.sh:30]: This is the overall workflow I'm imagining:

1. Run a bash/python script that iterates over all 40 salient regions and runs eedl to download images. It will also get images for every other region.
2. Run a python script that runs the image simulator and takes those images as input. The image simulator will vary the spacecraft position and orientation and pick random images from the dataset for each region to pull pixels from.
3. Train the RC and LD based on the training data generated by the image simulator.

So step 1 is going to be something like this script. How did you generate the dataset that you used to train the first LD network? Can we document that process in the form of some sort of script? 

Also, I'm working on step 2 based on this [these requirements](https://github.com/cmu-argus-2/GNC-Payload/issues/71), and I'll make sure it can interface with the existing code for step 3.
